In [98]:
import numpy as np
import pandas as pd
import datetime
%matplotlib inline
import matplotlib.pyplot as plt
import requests
import json

### First Way to Gather Data

In [99]:
twitter_arc = pd.read_csv(r'D:\Projects\twitter-archive-enhanced.csv')

In [100]:
twitter_arc.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193 NaN                    NaN                    
1  892177421306343426 NaN                    NaN                    
2  891815181378084864 NaN                    NaN                    
3  891689557279858688 NaN                    NaN                    
4  891327558926688256 NaN                    NaN                    

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3  This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0 NaN                  NaN                        NaN                         
1 NaN                  NaN                        NaN                         
2 NaN                  NaN                        NaN                         
3 NaN                  NaN                        NaN                         
4 NaN                  NaN                        NaN                         

                                                                                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1                                                                   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1                                                                   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1                                                                   
3  https://twitter.com/dog_rates/status/891689557279858688/photo/1                                                                   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   rating_numerator  rating_denominator      name doggo floofer pupper puppo  
0  13                10                  Phineas   None  None    None   None  
1  13                10                  Tilly     None  None    None   None  
2  12                10                  Archie    None  None    None   None  
3  13                10                  Darla     None  None    None   None  
4  12                10                  Franklin  None  None    None   None

### Second Way to Gather Data

In [101]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response= requests.get(url)

In [102]:
response= requests.get(url)

In [ ]:
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer

# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = twitter_arc.tweet_id.values
len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)


In [103]:
twitter_list = []
with open (r'D:\Projects\tweet-json.txt') as file:
    for line in file:
        twitter_list.append(json.loads(line))

In [104]:
print(twitter_list[0])

{'created_at': 'Tue Aug 01 16:23:56 +0000 2017', 'id': 892420643555336193, 'id_str': '892420643555336193', 'full_text': "This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU", 'truncated': False, 'display_text_range': [0, 85], 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 892420639486877696, 'id_str': '892420639486877696', 'indices': [86, 109], 'media_url': 'http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'url': 'https://t.co/MgUWQ76dJU', 'display_url': 'pic.twitter.com/MgUWQ76dJU', 'expanded_url': 'https://twitter.com/dog_rates/status/892420643555336193/photo/1', 'type': 'photo', 'sizes': {'large': {'w': 540, 'h': 528, 'resize': 'fit'}, 'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'small': {'w': 540, 'h': 528, 'resize': 'fit'}, 'medium': {'w': 540, 'h': 528, 'resize': 'fit'}}}]}, 'extended_entities': {'media

In [105]:
tweets_data = pd.DataFrame(twitter_list,columns=['id','retweet_count','favorite_count'])

In [106]:
tweets_data.head()

id  retweet_count  favorite_count
0  892420643555336193  8853           39467         
1  892177421306343426  6514           33819         
2  891815181378084864  4328           25461         
3  891689557279858688  8964           42908         
4  891327558926688256  9774           41048

In [107]:
tweets_data.rename(columns={'id':'tweet_id'},inplace=True)

In [108]:
tweets_data.to_csv('tweets_data.csv',index=False)

In [109]:
tweets_data = pd.read_csv('tweets_data.csv')

In [110]:
tweets_data.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193  8853           39467         
1  892177421306343426  6514           33819         
2  891815181378084864  4328           25461         
3  891689557279858688  8964           42908         
4  891327558926688256  9774           41048

### Third DataFrame

In [111]:
image_prediction = pd.read_csv(r'D:\Projects\image-predictions.tsv',sep='\t')

In [112]:
image_prediction.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0  1        Welsh_springer_spaniel  0.465074  True    collie               
1  1        redbone                 0.506826  True    miniature_pinscher   
2  1        German_shepherd         0.596461  True    malinois             
3  1        Rhodesian_ridgeback     0.408143  True    redbone              
4  1        miniature_pinscher      0.560311  True    Rottweiler           

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665  True    Shetland_sheepdog    0.061428  True    
1  0.074192  True    Rhodesian_ridgeback  0.072010  True    
2  0.138584  True    bloodhound           0.116197  True    
3  0.360687  True    miniature_pinscher   0.222752  True    
4  0.243682  True    Doberman             0.154629  True

## Assessing twitter_arc

In [113]:
twitter_arc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [114]:
twitter_arc.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1546  689283819090870273 NaN                    NaN                    
1981  672970152493887488 NaN                    NaN                    
374   828372645993398273 NaN                    NaN                    
1702  680934982542561280 NaN                    NaN                    
1451  695794761660297217 NaN                    NaN                    

                      timestamp  \
1546  2016-01-19 03:10:43 +0000   
1981  2015-12-05 02:46:02 +0000   
374   2017-02-05 22:40:03 +0000   
1702  2015-12-27 02:15:25 +0000   
1451  2016-02-06 02:22:53 +0000   

                                                                                  source  \
1546  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1981  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
374   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1702  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1451  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                     text  \
1546  Say hello to Chipson. He's aerodynamic af. No eyes (devastating). 9/10 would make sure he didn't bump into stuff https://t.co/V62rIva61J              
1981  This is Chet. He's having a hard time. Really struggling. 7/10 hang in there pupper https://t.co/eb4ta0xtnd                                           
374   This is Alexander Hamilpup. He was one of the many stars in this year's Puppy Bowl. He just hopes both teams had fun. 12/10 https://t.co/JcTuUcyYNS   
1702  Say hello to Sadie. She's a Tortellini Sidewinder. Very jubilant pup. Seems loyal. Leaves on point. 10/10 petable af https://t.co/g2bTu4ayPl          
1451  This is Wyatt. His throne is modeled after him. 13/10 Wyatt is a very big deal https://t.co/PccQ1CFEDd                                                

      retweeted_status_id  retweeted_status_user_id  \
1546 NaN                  NaN                         
1981 NaN                  NaN                         
374  NaN                  NaN                         
1702 NaN                  NaN                         
1451 NaN                  NaN                         

     retweeted_status_timestamp  \
1546  NaN                         
1981  NaN                         
374   NaN                         
1702  NaN                         
1451  NaN                         

                                                        expanded_urls  \
1546  https://twitter.com/dog_rates/status/689283819090870273/photo/1   
1981  https://twitter.com/dog_rates/status/672970152493887488/photo/1   
374   https://twitter.com/dog_rates/status/828372645993398273/photo/1   
1702  https://twitter.com/dog_rates/status/680934982542561280/photo/1   
1451  https://twitter.com/dog_rates/status/695794761660297217/photo/1   

      rating_numerator  rating_denominator       name doggo floofer  pupper  \
1546  9                 10                  Chipson    None  None    None     
1981  7                 10                  Chet       None  None    pupper   
374   12                10                  Alexander  None  None    None     
1702  10                10                  Sadie      None  None    None     
1451  13                10                  Wyatt      None  None    None     

     puppo  
1546  None  
1981  None  
374   None  
1702  None  
1451  None

In [115]:
twitter_arc.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03  7.800000e+01           7.800000e+01          
mean   7.427716e+17  7.455079e+17           2.014171e+16          
std    6.856705e+16  7.582492e+16           1.252797e+17          
min    6.660209e+17  6.658147e+17           1.185634e+07          
25%    6.783989e+17  6.757419e+17           3.086374e+08          
50%    7.196279e+17  7.038708e+17           4.196984e+09          
75%    7.993373e+17  8.257804e+17           4.196984e+09          
max    8.924206e+17  8.862664e+17           8.405479e+17          

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count  1.810000e+02         1.810000e+02              2356.000000        
mean   7.720400e+17         1.241698e+16              13.126486          
std    6.236928e+16         9.599254e+16              45.876648          
min    6.661041e+17         7.832140e+05              0.000000           
25%    7.186315e+17         4.196984e+09              10.000000          
50%    7.804657e+17         4.196984e+09              11.000000          
75%    8.203146e+17         4.196984e+09              12.000000          
max    8.874740e+17         7.874618e+17              1776.000000        

       rating_denominator  
count  2356.000000         
mean   10.455433           
std    6.745237            
min    0.000000            
25%    10.000000           
50%    10.000000           
75%    10.000000           
max    170.000000

In [116]:
twitter_arc['rating_denominator'].value_counts()

10     2333
11     3   
50     3   
20     2   
80     2   
0      1   
120    1   
7      1   
170    1   
150    1   
130    1   
90     1   
110    1   
2      1   
70     1   
40     1   
16     1   
15     1   
Name: rating_denominator, dtype: int64

In [117]:
twitter_arc['rating_numerator'].value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7       55 
14      54 
5       37 
6       32 
3       19 
4       17 
1       9  
2       9  
0       2  
15      2  
75      2  
420     2  
182     1  
204     1  
143     1  
121     1  
99      1  
20      1  
45      1  
27      1  
17      1  
24      1  
26      1  
44      1  
50      1  
60      1  
80      1  
84      1  
88      1  
1776    1  
960     1  
666     1  
144     1  
165     1  
Name: rating_numerator, dtype: int64

In [118]:
twitter_arc.query('rating_denominator==50')['rating_numerator']

1202    50
1274    45
1351    60
Name: rating_numerator, dtype: int64

In [119]:
twitter_arc[twitter_arc['retweeted_status_id'].notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257 NaN                    NaN                    
32    886054160059072513 NaN                    NaN                    
36    885311592912609280 NaN                    NaN                    
68    879130579576475649 NaN                    NaN                    
73    878404777348136964 NaN                    NaN                    
...                  ...  ..                     ..                    
1023  746521445350707200 NaN                    NaN                    
1043  743835915802583040 NaN                    NaN                    
1242  711998809858043904 NaN                    NaN                    
2259  667550904950915073 NaN                    NaN                    
2260  667550882905632768 NaN                    NaN                    

                      timestamp  \
19    2017-07-21 01:02:36 +0000   
32    2017-07-15 02:45:48 +0000   
36    2017-07-13 01:35:06 +0000   
68    2017-06-26 00:13:58 +0000   
73    2017-06-24 00:09:53 +0000   
...                         ...   
1023  2016-06-25 01:52:36 +0000   
1043  2016-06-17 16:01:16 +0000   
1242  2016-03-21 19:31:59 +0000   
2259  2015-11-20 03:51:52 +0000   
2260  2015-11-20 03:51:47 +0000   

                                                                                  source  \
19    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
32    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
36    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
68    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
73    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
...                                                                                  ...   
1023  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1043  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1242  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2259  <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                   
2260  <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                   

                                                                                                                                                text  \
19    RT @dog_rates: This is Canela. She attempted some fancy porch pics. They were unsuccessful. 13/10 someone help her https://t.co/cLyzpcUcMX       
32    RT @Athletics: 12/10 #BATP https://t.co/WxwJmvjfxo                                                                                               
36    RT @dog_rates: This is Lilly. She just parallel barked. Kindly requests a reward now. 13/10 would pet so well https://t.co/SATN4If5H5            
68    RT @dog_rates: This is Emmy. She was adopted today. Massive round of pupplause for Emmy and her new family. 14/10 for all involved https://…     
73    RT @dog_rates: Meet Shadow. In an attempt to reach maximum zooming borkdrive, he tore his ACL. Still 13/10 tho. Help him out below\n\nhttps:/…   
...                                                                                                                                              ...   
1023  RT @dog_rates: This is Shaggy. He knows exactly how to solve the puzzle but can't talk. All he wants to do is help. 10/10 great guy https:/…     
1043  RT @dog_rates: Extremely intelligent dog here. Has learned to walk like human. Even has his own dog. Very impressive 10/10 https://t.co/0Dv…     
1242  RT @twitter: @dog_rates Awesome Tweet! 12/10. Would Retweet. #LoveTwitter https://t.co/j6FQGhxYuN                                                
2259  RT @dogratingrating: Exceptional talent. Original humor. Cutting edge, Nova Scotian comedian. 12/10

In [120]:
twitter_arc[twitter_arc['retweeted_status_id'].notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257 NaN                    NaN                    
32    886054160059072513 NaN                    NaN                    
36    885311592912609280 NaN                    NaN                    
68    879130579576475649 NaN                    NaN                    
73    878404777348136964 NaN                    NaN                    
...                  ...  ..                     ..                    
1023  746521445350707200 NaN                    NaN                    
1043  743835915802583040 NaN                    NaN                    
1242  711998809858043904 NaN                    NaN                    
2259  667550904950915073 NaN                    NaN                    
2260  667550882905632768 NaN                    NaN                    

                      timestamp  \
19    2017-07-21 01:02:36 +0000   
32    2017-07-15 02:45:48 +0000   
36    2017-07-13 01:35:06 +0000   
68    2017-06-26 00:13:58 +0000   
73    2017-06-24 00:09:53 +0000   
...                         ...   
1023  2016-06-25 01:52:36 +0000   
1043  2016-06-17 16:01:16 +0000   
1242  2016-03-21 19:31:59 +0000   
2259  2015-11-20 03:51:52 +0000   
2260  2015-11-20 03:51:47 +0000   

                                                                                  source  \
19    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
32    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
36    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
68    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
73    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
...                                                                                  ...   
1023  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1043  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1242  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2259  <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                   
2260  <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                   

                                                                                                                                                text  \
19    RT @dog_rates: This is Canela. She attempted some fancy porch pics. They were unsuccessful. 13/10 someone help her https://t.co/cLyzpcUcMX       
32    RT @Athletics: 12/10 #BATP https://t.co/WxwJmvjfxo                                                                                               
36    RT @dog_rates: This is Lilly. She just parallel barked. Kindly requests a reward now. 13/10 would pet so well https://t.co/SATN4If5H5            
68    RT @dog_rates: This is Emmy. She was adopted today. Massive round of pupplause for Emmy and her new family. 14/10 for all involved https://…     
73    RT @dog_rates: Meet Shadow. In an attempt to reach maximum zooming borkdrive, he tore his ACL. Still 13/10 tho. Help him out below\n\nhttps:/…   
...                                                                                                                                              ...   
1023  RT @dog_rates: This is Shaggy. He knows exactly how to solve the puzzle but can't talk. All he wants to do is help. 10/10 great guy https:/…     
1043  RT @dog_rates: Extremely intelligent dog here. Has learned to walk like human. Even has his own dog. Very impressive 10/10 https://t.co/0Dv…     
1242  RT @twitter: @dog_rates Awesome Tweet! 12/10. Would Retweet. #LoveTwitter https://t.co/j6FQGhxYuN                                                
2259  RT @dogratingrating: Exceptional talent. Original humor. Cutting edge, Nova Scotian comedian. 12/10

In [121]:
twitter_arc.tail()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465 NaN                    NaN                    
2352  666044226329800704 NaN                    NaN                    
2353  666033412701032449 NaN                    NaN                    
2354  666029285002620928 NaN                    NaN                    
2355  666020888022790149 NaN                    NaN                    

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                                                  source  \
2351  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2352  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2353  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2354  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2355  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                             text  \
2351  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                      
2352  This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx     
2353  Here is a very happy pup. Big fan of well-maintained decks. Just look at that tongue. 9/10 would cuddle af https://t.co/y671yMhoiR            
2354  This is a western brown Mitsubishi terrier. Upset about leaf. Actually 2 dogs here. 7/10 would walk the shit out of https://t.co/r7mOb2m0UI   
2355  Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj           

      retweeted_status_id  retweeted_status_user_id  \
2351 NaN                  NaN                         
2352 NaN                  NaN                         
2353 NaN                  NaN                         
2354 NaN                  NaN                         
2355 NaN                  NaN                         

     retweeted_status_timestamp  \
2351  NaN                         
2352  NaN                         
2353  NaN                         
2354  NaN                         
2355  NaN                         

                                                        expanded_urls  \
2351  https://twitter.com/dog_rates/status/666049248165822465/photo/1   
2352  https://twitter.com/dog_rates/status/666044226329800704/photo/1   
2353  https://twitter.com/dog_rates/status/666033412701032449/photo/1   
2354  https://twitter.com/dog_rates/status/666029285002620928/photo/1   
2355  https://twitter.com/dog_rates/status/666020888022790149/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper puppo  
2351  5                 10                  None  None  None    None   None  
2352  6                 10                  a     None  None    None   None  
2353  9                 10                  a     None  None    None   None  
2354  7                 10                  a     None  None    None   None  
2355  8                 10                  None  None  None    None   None

In [122]:
twitter_arc['doggo'].value_counts()

None     2259
doggo    97  
Name: doggo, dtype: int64

In [123]:
twitter_arc['floofer'].value_counts()

None       2346
floofer    10  
Name: floofer, dtype: int64

In [124]:
twitter_arc['pupper'].value_counts()

None      2099
pupper    257 
Name: pupper, dtype: int64

In [125]:
twitter_arc['puppo'].value_counts()

None     2326
puppo    30  
Name: puppo, dtype: int64

In [126]:
twitter_arc['name'].value_counts()

None         745
a            55 
Charlie      12 
Oliver       11 
Lucy         11 
             .. 
Tove         1  
Tyrus        1  
Stephanus    1  
Brutus       1  
Smiley       1  
Name: name, Length: 957, dtype: int64

In [127]:
twitter_arc['source'].value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                        91  
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                     33  
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>    11  
Name: source, dtype: int64

In [128]:
pd.set_option("display.max_colwidth", -1)

<ipython-input-128-a09e5efec013>:1: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  pd.set_option("display.max_colwidth", -1)


In [129]:
twitter_arc[twitter_arc['in_reply_to_status_id'].notnull()][['text','rating_numerator','rating_denominator']]

text  \
30    @NonWhiteHat @MayhewMayhem omg hello tanner you are a scary good boy 12/10 would pet with extreme caution                                 
55    @roushfenway These are good dogs but 17/10 is an emotional impulse rating. More like 13/10s                                               
64    @RealKentMurphy 14/10 confirmed                                                                                                           
113   @ComplicitOwl @ShopWeRateDogs &gt;10/10 is reserved for dogs                                                                              
148   @Jack_Septic_Eye I'd need a few more pics to polish a full analysis, but based on the good boy content above I'm leaning towards 12/10    
...                                                                                                                                      ...    
2038  After 22 minutes of careful deliberation this dog is being demoted to a 1/10. The longer you look at him the more terrifying he becomes   
2149  After countless hours of research and hundreds of formula alterations we have concluded that Dug should be bumped to an 11/10             
2169  This is Tessa. She is also very pleased after finally meeting her biological father. 10/10 https://t.co/qDS1aCqppv                        
2189  12/10 good shit Bubka\n@wane15                                                                                                            
2298  After much debate this dog is being upgraded to 10/10. I repeat 10/10                                                                     

      rating_numerator  rating_denominator  
30    12                10                  
55    17                10                  
64    14                10                  
113   10                10                  
148   12                10                  
...   ..                ..                  
2038  1                 10                  
2149  11                10                  
2169  10                10                  
2189  12                10                  
2298  10                10                  

[78 rows x 3 columns]

In [130]:
len(twitter_arc.query('rating_denominator !=10'))

23

# Quality

#### we have 181 retweets to remove 
#### incorrect names like (a,None)
#### in_reply_to_status_id and in_reply_to_user_id is float 
#### datatype of timestamp is object
#### there is 23 rows in rating_denominator are not equal to 10 

# Tidy

#### we have to separate the time and date in timestamp column 
#### Dog stage is separated into four columns

# Second DataFrame

In [131]:
tweets_data.sample(5)

tweet_id  retweet_count  favorite_count
2318  666437273139982337  52             131           
2344  666058600524156928  61             115           
2245  667873844930215936  440            667           
1026  745789745784041472  1207           4437          
904   758041019896193024  433            3001

In [132]:
tweets_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2354 non-null   int64
 1   retweet_count   2354 non-null   int64
 2   favorite_count  2354 non-null   int64
dtypes: int64(3)
memory usage: 55.3 KB


In [133]:
tweets_data.isnull().any()

tweet_id          False
retweet_count     False
favorite_count    False
dtype: bool

In [134]:

tweets_data['retweet_count'].value_counts()

3652    5
1972    5
83      5
819     4
115     4
       ..
1526    1
230     1
1522    1
5616    1
7       1
Name: retweet_count, Length: 1724, dtype: int64

In [135]:
tweets_data['favorite_count'].value_counts()

0        179
2918     3  
2706     3  
2768     3  
1691     3  
        ..  
7493     1  
5444     1  
5442     1  
7489     1  
24574    1  
Name: favorite_count, Length: 2007, dtype: int64

In [136]:
tweets_data.describe()

tweet_id  retweet_count  favorite_count
count  2.354000e+03  2354.000000    2354.000000   
mean   7.426978e+17  3164.797366    8080.968564   
std    6.852812e+16  5284.770364    11814.771334  
min    6.660209e+17  0.000000       0.000000      
25%    6.783975e+17  624.500000     1415.000000   
50%    7.194596e+17  1473.500000    3603.500000   
75%    7.993058e+17  3652.000000    10122.250000  
max    8.924206e+17  79515.000000   132810.000000

# Tidy

### we could merge the two dataframes (twitter_arc , tweets_data) in one dataframe
# Quality
#### the count of tweet_id in twitter_arc is different from tweets_data
#### we have retweets to remove

# Third DataFrame

In [137]:
image_prediction.sample(5)

tweet_id                                          jpg_url  \
1139  729463711119904772  https://pbs.twimg.com/media/Ch-TXpFXAAAwPGf.jpg   
715   685547936038666240  https://pbs.twimg.com/media/CYOONfZW8AA7IOA.jpg   
920   701952816642965504  https://pbs.twimg.com/media/Cb3WXMUUMAIuzL8.jpg   
1004  708845821941387268  https://pbs.twimg.com/media/CdZTgynWwAATZcx.jpg   
1542  791312159183634433  https://pbs.twimg.com/media/CvtONV4WAAAQ3Rn.jpg   

      img_num                  p1   p1_conf  p1_dog                p2  \
1139  1        German_shepherd     0.829307  True    Doberman           
715   1        web_site            0.923987  False   oscilloscope       
920   1        toy_poodle          0.331707  True    miniature_poodle   
1004  1        schipperke          0.745640  True    kelpie             
1542  4        miniature_pinscher  0.892925  True    toy_terrier        

       p2_conf  p2_dog                  p3   p3_conf  p3_dog  
1139  0.022500  True    basenji             0.021190  True    
715   0.009712  False   hand-held_computer  0.008769  False   
920   0.272485  True    standard_poodle     0.169415  True    
1004  0.167853  True    Boston_bull         0.014763  True    
1542  0.095524  True    Doberman            0.003544  True

In [138]:
image_prediction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [139]:
image_prediction.columns[image_prediction.isnull().any()]

Index([], dtype='object')

In [140]:
image_prediction['img_num'].unique()

array([1, 4, 2, 3], dtype=int64)

In [141]:
image_prediction.query('p1_dog == True & p2_dog==True & p3_dog==True')

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2066  890609185150312448  https://pbs.twimg.com/media/DFwUU__XcAEpyXI.jpg   
2067  890729181411237888  https://pbs.twimg.com/media/DFyBahAVwAAhUTd.jpg   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0     1        Welsh_springer_spaniel  0.465074  True    collie               
1     1        redbone                 0.506826  True    miniature_pinscher   
2     1        German_shepherd         0.596461  True    malinois             
3     1        Rhodesian_ridgeback     0.408143  True    redbone              
4     1        miniature_pinscher      0.560311  True    Rottweiler           
...  ..                       ...           ...   ...           ...           
2066  1        Irish_terrier           0.487574  True    Irish_setter         
2067  2        Pomeranian              0.566142  True    Eskimo_dog           
2070  2        basset                  0.555712  True    English_springer     
2072  1        Chihuahua               0.716012  True    malamute             
2073  1        Chihuahua               0.323581  True    Pekinese             

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665  True    Shetland_sheepdog            0.061428  True    
1     0.074192  True    Rhodesian_ridgeback          0.072010  True    
2     0.138584  True    bloodhound                   0.116197  True    
3     0.360687  True    miniature_pinscher           0.222752  True    
4     0.243682  True    Doberman                     0.154629  True    
...        ...   ...         ...                          ...   ...    
2066  0.193054  True    Chesapeake_Bay_retriever     0.118184  True    
2067  0.178406  True    Pembroke                     0.076507  True    
2070  0.225770  True    German_short-haired_pointer  0.175219  True    
2072  0.078253  True    kelpie                       0.031379  True    
2073  0.090647  True    papillon                     0.068957  True    

[1243 rows x 12 columns]

# Quality
#### count of tweet_id is different from tweet_id in twitter_arc DataFrame
#### some names in P start lower letter and others with capital
#### names in P have _ instead of space
#### names in P have - instead of space
#### we have retweets to remove

# Tidy
#### we could merge the Three DataFrames in one DataFrame

# Clean

In [142]:
twitter_arc_clean = twitter_arc.copy()
tweets_data_clean = tweets_data.copy()
image_prediction_clean = image_prediction.copy()

# Define

### Drop retweets rows in the three DataFrames

### Code for first DataFrame

In [143]:
df = twitter_arc_clean[twitter_arc_clean['retweeted_status_id'].notnull()]
len(df)

181

In [144]:
twitter_arc_clean = twitter_arc_clean[~twitter_arc_clean['tweet_id'].isin(df['tweet_id'])]

### Test for first DataFrame

In [145]:
len(twitter_arc_clean)

2175

### Code for Second DataFrame

In [146]:
tweets_data_clean = tweets_data_clean[~tweets_data_clean['tweet_id'].isin(df['tweet_id'])]

### Test for second Dataframe

In [147]:
len(tweets_data_clean)

2175

In [1]:
# Now the count of tweets_data and twitter_arc is equal 

### Code for Third DataFrame

In [148]:
image_prediction_clean = image_prediction_clean[~image_prediction_clean['tweet_id'].isin(df['tweet_id'])]

### Test for Third DataFrame

In [149]:
len(image_prediction_clean)

1994

# Define

## Drop Columns of retweets 

### Code

In [151]:
twitter_arc_clean.drop(['retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'],axis=1,inplace=True)

### Test

In [152]:
len(twitter_arc_clean.columns)

14

In [153]:
#twitter_arc_clean[twitter_arc_clean['retweeted_status_id'].notnull()]

In [154]:
#twitter_arc_clean.drop(['retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'],axis=1,inplace=True)

In [155]:
#len(twitter_arc_clean.columns)

# Define

### Correct Data type for( 'in_reply_to_status_id' and 'in_reply_to_user_id' )

### Code 

In [156]:
twitter_arc_clean[['in_reply_to_status_id','in_reply_to_user_id']] = twitter_arc_clean[['in_reply_to_status_id','in_reply_to_user_id']].fillna(0)

In [157]:
twitter_arc_clean[['in_reply_to_status_id','in_reply_to_user_id']] = twitter_arc_clean[['in_reply_to_status_id','in_reply_to_user_id']].astype(int)

### Test

In [158]:
twitter_arc_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   tweet_id               2175 non-null   int64 
 1   in_reply_to_status_id  2175 non-null   int32 
 2   in_reply_to_user_id    2175 non-null   int32 
 3   timestamp              2175 non-null   object
 4   source                 2175 non-null   object
 5   text                   2175 non-null   object
 6   expanded_urls          2117 non-null   object
 7   rating_numerator       2175 non-null   int64 
 8   rating_denominator     2175 non-null   int64 
 9   name                   2175 non-null   object
 10  doggo                  2175 non-null   object
 11  floofer                2175 non-null   object
 12  pupper                 2175 non-null   object
 13  puppo                  2175 non-null   object
dtypes: int32(2), int64(3), object(9)
memory usage: 237.9+ KB


# Define

### Correct Data type for timestamp

### Code

In [159]:
twitter_arc_clean['timestamp'] = pd.to_datetime(twitter_arc_clean['timestamp'])

### Test

In [160]:
twitter_arc_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype              
---  ------                 --------------  -----              
 0   tweet_id               2175 non-null   int64              
 1   in_reply_to_status_id  2175 non-null   int32              
 2   in_reply_to_user_id    2175 non-null   int32              
 3   timestamp              2175 non-null   datetime64[ns, UTC]
 4   source                 2175 non-null   object             
 5   text                   2175 non-null   object             
 6   expanded_urls          2117 non-null   object             
 7   rating_numerator       2175 non-null   int64              
 8   rating_denominator     2175 non-null   int64              
 9   name                   2175 non-null   object             
 10  doggo                  2175 non-null   object             
 11  floofer                2175 non-null   object           

# Define

### Drop  23 rows in rating_denominator which are not equal to 10 

### Code

In [161]:
index_list = twitter_arc_clean.query('rating_denominator != 10').index
df1 = twitter_arc_clean.loc[index_list]

In [162]:
df1

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640 -2147483648             26259576              
342   832088576586297345 -2147483648             30582082              
433   820690176645140481  0                      0                     
516   810984652412424192  0                      0                     
902   758467244762497024  0                      0                     
1068  740373189193256964  0                      0                     
1120  731156023742988288  0                      0                     
1165  722974582966214656  0                      0                     
1202  716439118184652801  0                      0                     
1228  713900603437621249  0                      0                     
1254  710658690886586372  0                      0                     
1274  709198395643068416  0                      0                     
1351  704054845121142784  0                      0                     
1433  697463031882764288  0                      0                     
1598  686035780142297088 -2147483648            -2147483648            
1634  684225744407494656 -2147483648            -2147483648            
1635  684222868335505415  0                      0                     
1662  682962037429899265  0                      0                     
1663  682808988178739200 -2147483648            -2147483648            
1779  677716515794329600  0                      0                     
1843  675853064436391936  0                      0                     
2335  666287406224695296  0                      0                     

                     timestamp  \
313  2017-02-24 21:54:03+00:00   
342  2017-02-16 04:45:50+00:00   
433  2017-01-15 17:52:40+00:00   
516  2016-12-19 23:06:23+00:00   
902  2016-07-28 01:00:57+00:00   
1068 2016-06-08 02:41:38+00:00   
1120 2016-05-13 16:15:54+00:00   
1165 2016-04-21 02:25:47+00:00   
1202 2016-04-03 01:36:11+00:00   
1228 2016-03-27 01:29:02+00:00   
1254 2016-03-18 02:46:49+00:00   
1274 2016-03-14 02:04:08+00:00   
1351 2016-02-28 21:25:30+00:00   
1433 2016-02-10 16:51:59+00:00   
1598 2016-01-10 04:04:10+00:00   
1634 2016-01-05 04:11:44+00:00   
1635 2016-01-05 04:00:18+00:00   
1662 2016-01-01 16:30:13+00:00   
1663 2016-01-01 06:22:03+00:00   
1779 2015-12-18 05:06:23+00:00   
1843 2015-12-13 01:41:41+00:00   
2335 2015-11-16 16:11:11+00:00   

                                                                                  source  \
313   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
342   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
433   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
516   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
902   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1068  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1120  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1165  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1202  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1228  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1254  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1274  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1351  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1433  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1598  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1634  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1635  <a hre

In [163]:
twitter_arc_clean = twitter_arc_clean[~twitter_arc_clean['tweet_id'].isin(df1['tweet_id'])]

In [164]:
tweets_data_clean = tweets_data_clean[~tweets_data_clean['tweet_id'].isin(df1['tweet_id'])]

In [165]:
image_prediction_clean = image_prediction_clean[~image_prediction_clean['tweet_id'].isin(df1['tweet_id'])]

### Test

In [166]:
len(twitter_arc_clean)

2153

In [167]:
len(tweets_data_clean)

2153

In [168]:
len(image_prediction_clean)

1976

# Define

## Making the first letter capital in p1,p2,p3 columns 

### Code

In [169]:
image_prediction_clean['p1'] = image_prediction_clean['p1'].str.capitalize() 
image_prediction_clean['p2'] = image_prediction_clean['p2'].str.capitalize()
image_prediction_clean['p3'] = image_prediction_clean['p3'].str.capitalize()

### Test 

In [170]:
image_prediction_clean

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0     1        Welsh_springer_spaniel  0.465074  True    Collie               
1     1        Redbone                 0.506826  True    Miniature_pinscher   
2     1        German_shepherd         0.596461  True    Malinois             
3     1        Rhodesian_ridgeback     0.408143  True    Redbone              
4     1        Miniature_pinscher      0.560311  True    Rottweiler           
...  ..                       ...           ...   ...           ...           
2070  2        Basset                  0.555712  True    English_springer     
2071  1        Paper_towel             0.170278  False   Labrador_retriever   
2072  1        Chihuahua               0.716012  True    Malamute             
2073  1        Chihuahua               0.323581  True    Pekinese             
2074  1        Orange                  0.097049  False   Bagel                

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665  True    Shetland_sheepdog            0.061428  True    
1     0.074192  True    Rhodesian_ridgeback          0.072010  True    
2     0.138584  True    Bloodhound                   0.116197  True    
3     0.360687  True    Miniature_pinscher           0.222752  True    
4     0.243682  True    Doberman                     0.154629  True    
...        ...   ...         ...                          ...   ...    
2070  0.225770  True    German_short-haired_pointer  0.175219  True    
2071  0.168086  True    Spatula                      0.040836  False   
2072  0.078253  True    Kelpie                       0.031379  True    
2073  0.090647  True    Papillon                     0.068957  True    
2074  0.085851  False   Banana                       0.076110  False   

[1976 rows x 12 columns]

# Define

## Replacing _ with space in p1,p2,p3 columns

### Code

In [171]:
image_prediction_clean['p1'] = image_prediction_clean['p1'].str.replace('_',' ') 
image_prediction_clean['p2'] = image_prediction_clean['p2'].str.replace('_',' ') 
image_prediction_clean['p3'] = image_prediction_clean['p3'].str.replace('_',' ') 

### Test 

In [172]:
image_prediction_clean

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0     1        Welsh springer spaniel  0.465074  True    Collie               
1     1        Redbone                 0.506826  True    Miniature pinscher   
2     1        German shepherd         0.596461  True    Malinois             
3     1        Rhodesian ridgeback     0.408143  True    Redbone              
4     1        Miniature pinscher      0.560311  True    Rottweiler           
...  ..                       ...           ...   ...           ...           
2070  2        Basset                  0.555712  True    English springer     
2071  1        Paper towel             0.170278  False   Labrador retriever   
2072  1        Chihuahua               0.716012  True    Malamute             
2073  1        Chihuahua               0.323581  True    Pekinese             
2074  1        Orange                  0.097049  False   Bagel                

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665  True    Shetland sheepdog            0.061428  True    
1     0.074192  True    Rhodesian ridgeback          0.072010  True    
2     0.138584  True    Bloodhound                   0.116197  True    
3     0.360687  True    Miniature pinscher           0.222752  True    
4     0.243682  True    Doberman                     0.154629  True    
...        ...   ...         ...                          ...   ...    
2070  0.225770  True    German short-haired pointer  0.175219  True    
2071  0.168086  True    Spatula                      0.040836  False   
2072  0.078253  True    Kelpie                       0.031379  True    
2073  0.090647  True    Papillon                     0.068957  True    
2074  0.085851  False   Banana                       0.076110  False   

[1976 rows x 12 columns]

In [173]:
image_prediction_clean.select_dtypes([np.object]).apply(lambda x: x.str.contains('_').any())

jpg_url    True 
p1         False
p2         False
p3         False
dtype: bool

# Define 

## Replacing - with space in p1,p2,p3 columns

### Code 

In [174]:
image_prediction_clean['p1'] = image_prediction_clean['p1'].str.replace('-',' ') 
image_prediction_clean['p2'] = image_prediction_clean['p2'].str.replace('-',' ') 
image_prediction_clean['p3'] = image_prediction_clean['p3'].str.replace('-',' ') 

### Test

In [175]:
image_prediction_clean

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0     1        Welsh springer spaniel  0.465074  True    Collie               
1     1        Redbone                 0.506826  True    Miniature pinscher   
2     1        German shepherd         0.596461  True    Malinois             
3     1        Rhodesian ridgeback     0.408143  True    Redbone              
4     1        Miniature pinscher      0.560311  True    Rottweiler           
...  ..                       ...           ...   ...           ...           
2070  2        Basset                  0.555712  True    English springer     
2071  1        Paper towel             0.170278  False   Labrador retriever   
2072  1        Chihuahua               0.716012  True    Malamute             
2073  1        Chihuahua               0.323581  True    Pekinese             
2074  1        Orange                  0.097049  False   Bagel                

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665  True    Shetland sheepdog            0.061428  True    
1     0.074192  True    Rhodesian ridgeback          0.072010  True    
2     0.138584  True    Bloodhound                   0.116197  True    
3     0.360687  True    Miniature pinscher           0.222752  True    
4     0.243682  True    Doberman                     0.154629  True    
...        ...   ...         ...                          ...   ...    
2070  0.225770  True    German short haired pointer  0.175219  True    
2071  0.168086  True    Spatula                      0.040836  False   
2072  0.078253  True    Kelpie                       0.031379  True    
2073  0.090647  True    Papillon                     0.068957  True    
2074  0.085851  False   Banana                       0.076110  False   

[1976 rows x 12 columns]

In [176]:
image_prediction_clean.select_dtypes([np.object]).apply(lambda x: x.str.contains('_').any())

jpg_url    True 
p1         False
p2         False
p3         False
dtype: bool

# Tidy issues 

# Define

### Making dof_stage in one column instead of four columns

### Code

In [177]:
twitter_arc_clean['dog_stage'] = twitter_arc_clean['doggo']+twitter_arc_clean['floofer']+twitter_arc_clean['pupper']+twitter_arc_clean['puppo']

In [178]:
twitter_arc_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193  0                      0                     
1     892177421306343426  0                      0                     
2     891815181378084864  0                      0                     
3     891689557279858688  0                      0                     
4     891327558926688256  0                      0                     
...                  ... ..                     ..                     
2351  666049248165822465  0                      0                     
2352  666044226329800704  0                      0                     
2353  666033412701032449  0                      0                     
2354  666029285002620928  0                      0                     
2355  666020888022790149  0                      0                     

                     timestamp  \
0    2017-08-01 16:23:56+00:00   
1    2017-08-01 00:17:27+00:00   
2    2017-07-31 00:18:03+00:00   
3    2017-07-30 15:58:51+00:00   
4    2017-07-29 16:00:24+00:00   
...                        ...   
2351 2015-11-16 00:24:50+00:00   
2352 2015-11-16 00:04:52+00:00   
2353 2015-11-15 23:21:54+00:00   
2354 2015-11-15 23:05:30+00:00   
2355 2015-11-15 22:32:08+00:00   

                                                                                  source  \
0     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
...                                                                                  ...   
2351  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2352  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2353  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2354  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2355  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                    text  \
0     This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1     This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2     This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3     This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4     This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   
...                                                                                                                                                                  ...   
2351  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                                             
2352  This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx                            
2353  Here is a very happy pup. Big fan of well-maintained decks. Just look at that tongu

In [179]:
twitter_arc_clean['dog_stage'].value_counts()

NoneNoneNoneNone        1809
NoneNonepupperNone      224 
doggoNoneNoneNone       75  
NoneNoneNonepuppo       24  
doggoNonepupperNone     10  
NoneflooferNoneNone     9   
doggoflooferNoneNone    1   
doggoNoneNonepuppo      1   
Name: dog_stage, dtype: int64

In [180]:
for i in range(len(twitter_arc_clean)):
    if twitter_arc_clean.iloc[i,14]=='NoneNoneNoneNone':
        twitter_arc_clean.iloc[i,14] = 'None'
    elif twitter_arc_clean.iloc[i,14]=='doggoNoneNoneNone':
        twitter_arc_clean.iloc[i,14] = 'doggo'
    elif twitter_arc_clean.iloc[i,14]=='NoneflooferNoneNone':
        twitter_arc_clean.iloc[i,14] = 'floofer'
    elif twitter_arc_clean.iloc[i,14]=='NoneNonepupperNone':
        twitter_arc_clean.iloc[i,14] = 'pupper'
    elif twitter_arc_clean.iloc[i,14]=='NoneNoneNonepuppo':
        twitter_arc_clean.iloc[i,14] = 'puppo'
    elif twitter_arc_clean.iloc[i,14]== 'doggoNonepupperNone':
        twitter_arc_clean.iloc[i,14]= 'doggo or pupper'
    elif twitter_arc_clean.iloc[i,14]== 'doggoflooferNoneNone':
        twitter_arc_clean.iloc[i,14]= 'doggo or floofer'
    elif twitter_arc_clean.iloc[i,14]== 'doggoNoneNonepuppo':
        twitter_arc_clean.iloc[i,14]= 'doggo or puppo'

### Test

In [181]:
twitter_arc_clean['dog_stage'].value_counts()

None                1809
pupper              224 
doggo               75  
puppo               24  
doggo or pupper     10  
floofer             9   
doggo or floofer    1   
doggo or puppo      1   
Name: dog_stage, dtype: int64

In [182]:
twitter_arc_clean.drop(['pupper','doggo','puppo','floofer'],axis=1,inplace=True)

In [183]:
twitter_arc_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193  0                      0                     
1     892177421306343426  0                      0                     
2     891815181378084864  0                      0                     
3     891689557279858688  0                      0                     
4     891327558926688256  0                      0                     
...                  ... ..                     ..                     
2351  666049248165822465  0                      0                     
2352  666044226329800704  0                      0                     
2353  666033412701032449  0                      0                     
2354  666029285002620928  0                      0                     
2355  666020888022790149  0                      0                     

                     timestamp  \
0    2017-08-01 16:23:56+00:00   
1    2017-08-01 00:17:27+00:00   
2    2017-07-31 00:18:03+00:00   
3    2017-07-30 15:58:51+00:00   
4    2017-07-29 16:00:24+00:00   
...                        ...   
2351 2015-11-16 00:24:50+00:00   
2352 2015-11-16 00:04:52+00:00   
2353 2015-11-15 23:21:54+00:00   
2354 2015-11-15 23:05:30+00:00   
2355 2015-11-15 22:32:08+00:00   

                                                                                  source  \
0     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
...                                                                                  ...   
2351  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2352  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2353  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2354  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2355  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                    text  \
0     This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1     This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2     This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3     This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4     This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   
...                                                                                                                                                                  ...   
2351  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                                             
2352  This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx                            
2353  Here is a very happy pup. Big fan of well-maintained decks. Just look at that tongu

In [184]:
twitter_arc_clean['dog_stage'] = twitter_arc_clean['dog_stage'].astype('category')

In [185]:
twitter_arc_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2153 entries, 0 to 2355
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype              
---  ------                 --------------  -----              
 0   tweet_id               2153 non-null   int64              
 1   in_reply_to_status_id  2153 non-null   int32              
 2   in_reply_to_user_id    2153 non-null   int32              
 3   timestamp              2153 non-null   datetime64[ns, UTC]
 4   source                 2153 non-null   object             
 5   text                   2153 non-null   object             
 6   expanded_urls          2099 non-null   object             
 7   rating_numerator       2153 non-null   int64              
 8   rating_denominator     2153 non-null   int64              
 9   name                   2153 non-null   object             
 10  dog_stage              2153 non-null   category           
dtypes: category(1), datetime64[ns, UTC](1), int32(2), int64(

# Define

## Merging the three DataFrames in one DataFrames

### Code

In [186]:
df_final = pd.merge(twitter_arc_clean,tweets_data_clean,on=['tweet_id'],how='outer')

In [187]:
df_final

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193  0                      0                     
1     892177421306343426  0                      0                     
2     891815181378084864  0                      0                     
3     891689557279858688  0                      0                     
4     891327558926688256  0                      0                     
...                  ... ..                     ..                     
2148  666049248165822465  0                      0                     
2149  666044226329800704  0                      0                     
2150  666033412701032449  0                      0                     
2151  666029285002620928  0                      0                     
2152  666020888022790149  0                      0                     

                     timestamp  \
0    2017-08-01 16:23:56+00:00   
1    2017-08-01 00:17:27+00:00   
2    2017-07-31 00:18:03+00:00   
3    2017-07-30 15:58:51+00:00   
4    2017-07-29 16:00:24+00:00   
...                        ...   
2148 2015-11-16 00:24:50+00:00   
2149 2015-11-16 00:04:52+00:00   
2150 2015-11-15 23:21:54+00:00   
2151 2015-11-15 23:05:30+00:00   
2152 2015-11-15 22:32:08+00:00   

                                                                                  source  \
0     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
...                                                                                  ...   
2148  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2149  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2150  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2151  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2152  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                    text  \
0     This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1     This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2     This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3     This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4     This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   
...                                                                                                                                                                  ...   
2148  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                                             
2149  This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx                            
2150  Here is a very happy pup. Big fan of well-maintained decks. Just look at that tongu

In [188]:
df_final = pd.merge(df_final,image_prediction_clean,on=['tweet_id'],how='outer')

### Test

In [189]:
df_final

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193  0                      0                     
1     892177421306343426  0                      0                     
2     891815181378084864  0                      0                     
3     891689557279858688  0                      0                     
4     891327558926688256  0                      0                     
...                  ... ..                     ..                     
2148  666049248165822465  0                      0                     
2149  666044226329800704  0                      0                     
2150  666033412701032449  0                      0                     
2151  666029285002620928  0                      0                     
2152  666020888022790149  0                      0                     

                     timestamp  \
0    2017-08-01 16:23:56+00:00   
1    2017-08-01 00:17:27+00:00   
2    2017-07-31 00:18:03+00:00   
3    2017-07-30 15:58:51+00:00   
4    2017-07-29 16:00:24+00:00   
...                        ...   
2148 2015-11-16 00:24:50+00:00   
2149 2015-11-16 00:04:52+00:00   
2150 2015-11-15 23:21:54+00:00   
2151 2015-11-15 23:05:30+00:00   
2152 2015-11-15 22:32:08+00:00   

                                                                                  source  \
0     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4     <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
...                                                                                  ...   
2148  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2149  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2150  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2151  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2152  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                    text  \
0     This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1     This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2     This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3     This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4     This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   
...                                                                                                                                                                  ...   
2148  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                                             
2149  This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx                            
2150  Here is a very happy pup. Big fan of well-maintained decks. Just look at that tongu

In [190]:
df_final.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'dog_stage', 'retweet_count',
       'favorite_count', 'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2',
       'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'],
      dtype='object')

In [191]:
df_final.to_csv('twitter_archive_master.csv',index=False)